src/masternodeman.cpp

```diff
@@ -176,30 +175,103 @@ void CMasternodeMan::CheckAndRemove()
+                    // this mn is in a non-recoverable state and we haven't asked other nodes yet
+                    std::set<CNetAddr> setRequested;
+                    // calulate only once and only when it's needed
+                    if(vecMasternodeRanks.empty()) {
+                        int nRandomBlockHeight = GetRandInt(pCurrentBlockIndex->nHeight);
+                        vecMasternodeRanks = GetMasternodeRanks(nRandomBlockHeight);
+                    }
+                    // ask first MNB_RECOVERY_QUORUM_TOTAL mns we can connect to and we haven't asked recently
+                    for(int i = 0; setRequested.size() < MNB_RECOVERY_QUORUM_TOTAL && i < (int)vecMasternodeRanks.size(); i++) {
+                        // avoid banning
+                        if(mWeAskedForMasternodeListEntry.count(it->vin.prevout) && mWeAskedForMasternodeListEntry[it->vin.prevout].count(vecMasternodeRanks[i].second.addr)) continue;
+                        // didn't ask recently, ok to ask now
+                        CService addr = vecMasternodeRanks[i].second.addr;
+                        CNode* pnode = ConnectNode(CAddress(addr), NULL, true);
+                        if(pnode) {
+                            LogPrint("masternode", "CMasternodeMan::CheckAndRemove -- asking for mnb of %s, addr=%s\n", it->vin.prevout.ToStringShort(), addr.ToString());
+                            setRequested.insert(addr);
+                            // can't use AskForMN here, inv system is way too smart, request data directly instead
+                            std::vector<CInv> vToFetch;
+                            vToFetch.push_back(CInv(MSG_MASTERNODE_ANNOUNCE, hash));
+                            pnode->PushMessage(NetMsgType::GETDATA, vToFetch);
+                            fAskedForMnbRecovery = true;
+                        } else {
+                            LogPrint("masternode", "CMasternodeMan::CheckAndRemove -- can't connect to node to ask for mnb, addr=%s\n", addr.ToString());
+                        }
+                    }
+                    // wait for mnb recovery replies for MNB_RECOVERY_WAIT_SECONDS seconds
+                    mMnbRecoveryRequests[hash] = std::make_pair(GetTime() + MNB_RECOVERY_WAIT_SECONDS, setRequested);
+                }
                 ++it;
             }
         }
 
+        // proces replies for MASTERNODE_NEW_START_REQUIRED masternodes
+        LogPrint("masternode", "CMasternodeMan::CheckAndRemove -- mMnbRecoveryGoodReplies size=%d\n", (int)mMnbRecoveryGoodReplies.size());
+        std::map<uint256, std::vector<CMasternodeBroadcast> >::iterator itMnbReplies = mMnbRecoveryGoodReplies.begin();
+        while(itMnbReplies != mMnbRecoveryGoodReplies.end()){
+            if(mMnbRecoveryRequests[itMnbReplies->first].first < GetTime()) {
+                // all nodes we asked should have replied now
+                if(itMnbReplies->second.size() > MNB_RECOVERY_QUORUM_REQUIRED) {
```

Comment:
if you mean `itMnbReplies->second.size() >= MNB_RECOVERY_QUORUM_REQUIRED` - yep, good catch, will fix!